src/id_allocator.rs

```diff
@@ -0,0 +1,182 @@
+//! that can be abstracted to an integer.
+
+use crate::Error;
+use crate::Result;
+use std::collections::BTreeSet;
+
+/// Internal representation of IdAllocator. Contains the ends of the interval
+/// that is managed, a field that points at the next available ID, and a
+/// BTreeSet used to store the released IDs. The reason we have chosed a
+/// BTreeSet is that the average complexity for deletion and insertion is
+/// O(logN) compared to Vec for example, another benefit is that the entries
+/// are sorted so we will always use the first available ID.
+#[derive(Debug)]
+pub struct IdAllocator {
+    // Beginnbing of the range of IDs that we want to manage.
+    range_base: u32,
+    // First available id that was never allocated.
+    next_id: Option<u32>,
+    // End of the range of IDs that we want to manage.
+    range_end: u32,
+    // Set of all freed ids that can be reused at subsequent allocations.
+    freed_ids: BTreeSet<u32>,
+}
+
+impl IdAllocator {
+    /// Creates a new instance of IdAllocator that will be used to manage the
+    /// allocation and release of ids from the interval specified by
+    /// `range_base` and `range_end`
+    pub fn new(range_base: u32, range_end: u32) -> std::result::Result<Self, Error> {
+        if range_end < range_base {
+            return Err(Error::InvalidRange(range_base, range_end));
+        }
+        Ok(IdAllocator {
+            range_base,
+            next_id: Some(range_base),
+            range_end,
+            freed_ids: BTreeSet::<u32>::new(),
+        })
+    }
+
+    fn id_in_range(&self, id: u32) -> bool {
+        // Check for out of range.
+        self.range_base <= id && id <= self.range_end
+    }
+
+    /// Allocate an ID from the managed ranged.
+    /// We first try to reuse one of the IDs released before. If there is no
+    /// ID to reuse we return the next available one from the managed range.
+    pub fn allocate_id(&mut self) -> Result {
+        // If the set containing all freed ids is not empty we extract the
+        // first entry from that set and return it.
+        if !self.freed_ids.is_empty() {
+            let ret_value = *self.freed_ids.iter().next().unwrap();
+            self.freed_ids.remove(&ret_value);
+            return Ok(ret_value);
+        }
+        // If no id was freed before we return the next available id.
+        if self.next_id.is_none() {
+            return Err(Error::Overflow);
+        }
+        // Unwrap is safe because we already checked that next_id is not None.
+        let id = self.next_id.unwrap();
+        if id > self.range_end {
+            return Err(Error::ResourceExhausted);
+        }
+        // Prepare the next available id and return `Error::Overflow` if the
+        // addition results in an overflow.
+        // Unwrap is safe because we already checked that next_id is not None.
+        self.next_id = self.next_id.unwrap().checked_add(1);
```

Comment:
If we use a `match` or `if let` above, we can use a `map` here to avoid the `unwrap`.